x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/rest/StreamingRestChunkedToXContentListener.java

```diff
@@ -0,0 +1,460 @@
+ * the response in Server-Sent Events.
+ */
+public class StreamingRestChunkedToXContentListener implements ActionListener<InferenceAction.Response> {
+    private static final Logger logger = LogManager.getLogger(StreamingRestChunkedToXContentListener.class);
+    private final StreamingSubscriber subscriber = new StreamingSubscriber();
+    private final AtomicBoolean isLastPart = new AtomicBoolean(false);
+    private final RestChannel channel;
+    private final ToXContent.Params params;
+
+    /**
+     * A listener for the first part of the next entry to become available for transmission.
+     * Chunks are sent one at a time through the completion of this listener.
+     * This listener is initialized in {@link #initializeStream(InferenceAction.Response)} before the first chunk is requested.
+     * When a chunk is ready, this listener is completed with the converted {@link ChunkedRestResponseBodyPart}.
+     * After transmitting the chunk, {@link #requestNextChunk(ActionListener)} will set the next listener and request the next
+     * chunk. This cycle will repeat until this listener completes with the DONE chunk and the stream closes.
+     */
+    private SubscribableListener<ChunkedRestResponseBodyPart> nextBodyPartListener;
+
+    public StreamingRestChunkedToXContentListener(RestChannel channel) {
+        this(channel, channel.request());
+    }
+
+    public StreamingRestChunkedToXContentListener(RestChannel channel, ToXContent.Params params) {
+        this.channel = channel;
+        this.params = params;
+    }
+
+    @Override
+    public void onResponse(InferenceAction.Response response) {
+        try {
+            ensureOpen();
+            if (response.isStreaming()) {
+                initializeStream(response);
+            } else {
+                // if we aren't streaming - if the provider doesn't allow streaming, there will be a single message
+                channel.sendResponse(
+                    RestResponse.chunked(RestStatus.OK, new SingleServerSentEventBodyPart(ServerSentEvents.MESSAGE, response), () -> {})
+                );
+            }
+        } catch (Exception e) {
+            onFailure(e);
+        }
+    }
+
+    protected void ensureOpen() {
+        if (channel.request().getHttpChannel().isOpen() == false) {
+            onFailure(new TaskCancelledException("response channel [" + channel.request().getHttpChannel() + "] closed"));
+        }
+    }
+
+    private void initializeStream(InferenceAction.Response response) {
+        nextBodyPartListener = new SubscribableListener<>();
+        nextBodyPartListener.addListener(ActionListener.wrap(bodyPart -> {
+            // this is the first response, so we need to send the RestResponse to open the stream
+            // all subsequent bytes will be delivered through the nextBodyPartListener
+            channel.sendResponse(RestResponse.chunked(RestStatus.OK, bodyPart, this::release));
+        }, this::onFailure));
```

Comment:
I'm suspicious about subscribing `this::onFailure` here. This is called once this listener is completed normally, we shouldn't be completing it again exceptionally.